cpp/src/arrow/memory_pool.cc

```diff
@@ -83,10 +87,94 @@ const char* je_arrow_malloc_conf =
 #endif  // ARROW_JEMALLOC
 
 namespace arrow {
+namespace {
 
 constexpr size_t kAlignment = 64;
 
-namespace {
+constexpr char kDefaultBackendEnvVar[] = "ARROW_DEFAULT_MEMORY_POOL";
+
+enum class MemoryPoolBackend : uint8_t { System, Jemalloc, Mimalloc };
+
+std::string BackendName(MemoryPoolBackend backend) {
+  switch (backend) {
+    case MemoryPoolBackend::System:
+      return "system";
+    case MemoryPoolBackend::Jemalloc:
+      return "jemalloc";
+    case MemoryPoolBackend::Mimalloc:
+      return "mimalloc";
+    default:
+      return "";  // XXX unreachable
+  }
+}
+
+std::vector<MemoryPoolBackend> SupportedBackends() {
+  std::vector<MemoryPoolBackend> backends = {
+#ifdef ARROW_JEMALLOC
+      MemoryPoolBackend::Jemalloc,
+#endif
+#ifdef ARROW_MIMALLOC
+      MemoryPoolBackend::Mimalloc,
+#endif
+      MemoryPoolBackend::System};
+  return backends;
+}
+
+const std::vector<MemoryPoolBackend> supported_backends = SupportedBackends();
+
+util::optional<MemoryPoolBackend> UserSelectedBackend() {
+  auto unsupported_backend = [](const std::string& name) {
+    std::vector<std::string> supported;
+    for (const auto backend : supported_backends) {
+      supported.push_back("'" + BackendName(backend) + "'");
+    }
+    ARROW_LOG(WARNING) << "Unsupported backend '" << name << "' specified in "
+                       << kDefaultBackendEnvVar << " (supported backends are "
+                       << internal::JoinStrings(supported, ", ") << ")";
+  };
+
+  auto maybe_name = internal::GetEnvVar(kDefaultBackendEnvVar);
+  if (!maybe_name.ok()) {
+    return {};
+  }
+  const auto name = *std::move(maybe_name);
+  if (name.empty()) {
+    return {};
+  } else if (name == "system") {
+    return MemoryPoolBackend::System;
+  } else if (name == "jemalloc") {
+#ifdef ARROW_JEMALLOC
+    return MemoryPoolBackend::Jemalloc;
+#else
+    unsupported_backend(name);
+#endif
+  } else if (name == "mimalloc") {
+#ifdef ARROW_MIMALLOC
+    return MemoryPoolBackend::Mimalloc;
+#else
+    unsupported_backend(name);
+#endif
+  } else {
+    unsupported_backend(name);
+  }
```

Comment:
Nit: this would be more readable if you iterated through SupportedBackends and created a `map<name, backend>` then did lookup in that